File: Day7/solution.js
const fs = require('fs');
const data = fs.readFileSync('./input.txt', 'utf-8').split('\r\n')

const bagRules = {}

for (let i = 0; i < data.length; i++) {
    let [bag1, ...bags] = data[i].split('contain');
    bag1 = bag1.replace(/[0-9.]|\sbags?.?[^a-zA-Z]/g, '')

    let contains = bags[0].split(',').map(e => {
        if (e.match(/no other bags/)) {
            return null
        } else {
            return {
                num: parseInt(e.match(/\d+/)),
                type: e.replace(/[0-9.]/g, '').replace(/bags?/g,'').trim()
            }
        }

    })



    bagRules[bag1] = contains[0] === null ? null : contains
}




function checkContains(rule, target) {
    if (bagRules[rule] === null) {
        return false
    }

    let typesInside = bagRules[rule].map(e => {
        return e.type
    })

    if (typesInside.includes(target)) {
        return true
    } else {
        let isAllowed = typesInside.filter(e => {
            return checkContains(e, target)
        })

        return isAllowed.length > 0 ? true : false
    }
}

let part1 = Object.keys(bagRules).filter(e => {
    return checkContains(e, 'shiny gold')
})
console.log('Part 1: ', part1.length);


// Part 2:
function checkNumber(type) {
    if (bagRules[type] === null) {
        return 0
    } 
    
    let count = 0
    for (bag of bagRules[type]) {
        count += bag.num
        count += bag.num * checkNumber(bag.type)
    }

    return count

}

console.log('Part 2: ', checkNumber('shiny gold'))

File: Day7/solution.py
import re

data = open('./input.txt').read().splitlines()

bag_rules = {}

for line in data:
    bag, bags = line.split('contain')
    bag = re.match(r'(\w+ \w+)', bag).group()
    bags_inside = re.findall(r'(\d+ \w+ \w+) bags?', bags)
    bag_rules[bag] = list()

    if bags_inside:
        for b in bags_inside:
            num, bn1, bn2 = b.split(' ')
            bagname = f'{bn1} {bn2}'
            bag_rules[bag].append({'type': bagname, 'num': int(num)})


def check_inside(bag, target):
    if len(bag_rules[bag]) == 0:
        return False

    target_inside = False

    for b in bag_rules[bag]:
        if b['type'] == target:
            target_inside = True
            break
        elif check_inside(b['type'], target) == True:
            target_inside = True

    return target_inside


part1 = 0
for bag in bag_rules.keys():
    if check_inside(bag, 'shiny gold'):
        part1 += 1

print(part1)
